src/Services/AuthenticationService.php

```diff
@@ -111,23 +134,57 @@ public function getAuthenticateUser(
         $userId = (string)$claims[$this->userIdAttr][0];
         $user = $this->userRepository->getUserEntityByIdentifier($userId);
 
-        if (!$user) {
-            $user = UserEntity::fromData($userId, $claims);
-            $this->userRepository->add($user);
-        } else {
+        if ($user) {
             $user->setClaims($claims);
             $this->userRepository->update($user);
+        } else {
+            $user = UserEntity::fromData($userId, $claims);
+            $this->userRepository->add($user);
+        }
+
+        if (empty($state['Oidc']['RelyingPartyMetadata']['id'])) {
+            throw new Error\Exception('OIDC RelyingPartyMetadata ID does not exist in state');
         }
 
-        $this->addRelyingPartyAssociation($oidcClient, $user);
+        $client = $this->clientRepository->findById((string)$state['Oidc']['RelyingPartyMetadata']['id']);
+        if (!$client) {
+            throw new Error\NotFound('Client not found.');
+        }
+
+        $this->addRelyingPartyAssociation($client, $user);
 
         return $user;
     }
 
+    /**
+     * @param   array  $state
+     *
+     * @return AuthorizationRequest
+     * @throws Exception
+     */
+
+    public function getAuthorizationRequestFromState(array $state): AuthorizationRequest
+    {
+        if (!($state['authorizationRequest'] instanceof AuthorizationRequest)) {
+            throw new Exception("Authorization Request is not valid: " . print_r($state, true));
```

Comment:
I don't think the `$state` value should appear in the exception. I think there is often data in the `$state` array that we don't want to appear in exception messages.  For example we (Cirrus) don't want a user's personal information to appear in our logs, and that PII is in the `$state` array and will be part of the exception message.

Reply:
done